File: android/core/data/src/main/kotlin/com/riskdetectedan/core/data/notebook/NotebookReader.kt
package com.riskdetectedan.core.data.notebook

import kotlinx.coroutines.currentCoroutineContext
import kotlinx.coroutines.ensureActive
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import java.util.UUID

@Serializable
data class NotebookRecord(val note_id: String, val title: String?, val body: String?, val version: Long,
    val tombstone: Boolean, val updated_at: String) {
    fun validate() {
        require(UUID.fromString(note_id).toString() == note_id && version in 1..9007199254740991L && updated_at.isNotEmpty())
        require((title?.codePointCount(0, title.length) ?: 0) <= 200 && (body?.codePointCount(0, body.length) ?: 0) <= 20000)
        require(!tombstone || (title == null && body == null))
    }
}

/** Session-only read model; durable edits stay in the separate encrypted outbox. */
class NotebookReader(private val current: () -> NotebookIdentity?, private val read: suspend (String?) -> String) {
    data class Snapshot(val notes: List<NotebookRecord>, val drafts: List<NotebookPending>)
    @Serializable private data class Page(val schema_version: Int, val scan_mode: String, val notes: List<NotebookRecord>,
        val has_more: Boolean, val next_after: String?)
    private val lock = Mutex()
    private var bound: NotebookIdentity? = null
    private var records = mapOf<String, NotebookRecord>()
    private suspend fun check(identity: NotebookIdentity) {
        currentCoroutineContext().ensureActive()
        if (current() != identity) { records = emptyMap(); bound = null; throw NotebookFailure("IDENTITY_CHANGED") }
        if (bound != identity) { records = emptyMap(); bound = identity }
    }
    suspend fun snapshot(identity: NotebookIdentity, drafts: List<NotebookPending>): Snapshot = lock.withLock {
        check(identity); Snapshot(records.values.filterNot { it.tombstone }.sortedBy { it.note_id }, drafts)
    }
    suspend fun refresh(identity: NotebookIdentity) = lock.withLock {
        check(identity)
        var after: String? = null
        val merged = records.toMutableMap()
        repeat(50) {
            val raw = try { read(after) } catch (error: Exception) { check(identity); throw NotebookFailure("UNAVAILABLE") }
            check(identity); require(raw.toByteArray(Charsets.UTF_8).size <= 2_000_000)
            val page = Json.decodeFromString<Page>(raw)
            require(page.schema_version == 1 && page.scan_mode == "full_scan_restart" && page.notes.size <= 20)
            require(if (page.has_more) page.notes.size == 20 && page.next_after == page.notes.last().note_id else page.next_after == null)
            var previous = after ?: ""
            for (record in page.notes) {
                record.validate(); require(record.note_id > previous); previous = record.note_id
                val old = merged[record.note_id]
                if (old != null) {
                    if (old.tombstone || record.version < old.version) continue
                    if (old.version == record.version) { require(old == record); continue }
                }
                merged[record.note_id] = record
            }
            if (merged.size > 1000 || Json.encodeToString(merged.values.toList()).toByteArray(Charsets.UTF_8).size > 2_000_000) throw NotebookFailure("QUEUE_FULL")
            if (!page.has_more) { records = merged; return@withLock }
            after = page.next_after
        }
        throw NotebookFailure("QUEUE_FULL")
    }
}

File: App/Services/Notebook/NotebookReader.swift
import Foundation

struct NotebookRecord: Codable, Equatable {
    let note_id: UUID
    let title: String?
    let body: String?
    let version: Int64
    let tombstone: Bool
    let updated_at: String
    func validate() throws {
        guard (1...9007199254740991).contains(version), !updated_at.isEmpty,
              (title?.unicodeScalars.count ?? 0) <= 200, (body?.unicodeScalars.count ?? 0) <= 20000,
              !tombstone || (title == nil && body == nil) else { throw NotebookFailure.invalid }
    }
}

/// Session-bound in-memory read model. Durable drafts remain in the separate outbox.
/// Missing rows never delete data; old responses never resurrect a tombstone.
@MainActor final class NotebookReader {
    struct Snapshot { let notes: [NotebookRecord]; let drafts: [NotebookPending] }
    private struct Page: Decodable {
        let schema_version: Int; let scan_mode: String; let notes: [NotebookRecord]
        let has_more: Bool; let next_after: UUID?
    }
    private let current: () -> NotebookIdentity?
    private let read: (UUID?) async throws -> Data
    private var bound: NotebookIdentity?
    private var records = [UUID: NotebookRecord]()
    private var fetching = false
    init(current: @escaping () -> NotebookIdentity?, read: @escaping (UUID?) async throws -> Data) {
        self.current = current; self.read = read
    }
    private func check(_ identity: NotebookIdentity) throws {
        try Task.checkCancellation()
        guard current() == identity else {
            records.removeAll(); bound = nil; throw NotebookFailure.identityChanged
        }
        if bound != identity { records.removeAll(); bound = identity }
    }
    func snapshot(_ identity: NotebookIdentity, drafts: [NotebookPending]) throws -> Snapshot {
        try check(identity)
        // Keep pending edits separate, including edits blocked by a tombstone.
        return Snapshot(notes: records.values.filter { !$0.tombstone }.sorted { $0.note_id.uuidString < $1.note_id.uuidString }, drafts: drafts)
    }
    func refresh(_ identity: NotebookIdentity) async throws {
        try check(identity)
        guard !fetching else { throw NotebookFailure.busy }
        fetching = true; defer { fetching = false }
        var after: UUID? = nil
        // Bounded full scan; only publish after every page passes validation.
        var merged = records
        for _ in 0..<50 {
            let data: Data
            do { data = try await read(after) } catch { try check(identity); throw NotebookFailure.unavailable }
            try check(identity)
            guard data.count <= 2_000_000 else { throw NotebookFailure.invalid }
            let page = try JSONDecoder().decode(Page.self, from: data)
            guard page.schema_version == 1, page.scan_mode == "full_scan_restart", page.notes.count <= 20,
                  page.has_more ? page.notes.count == 20 && page.next_after == page.notes.last?.note_id : page.next_after == nil else { throw NotebookFailure.invalid }
            var previous = after?.uuidString ?? ""
            for record in page.notes {
                try record.validate()
                guard record.note_id.uuidString > previous else { throw NotebookFailure.invalid }
                previous = record.note_id.uuidString
                if let old = merged[record.note_id] {
                    if old.tombstone || record.version < old.version { continue }
                    if record.version == old.version {
                        guard record == old else { throw NotebookFailure.invalid }; continue
                    }
                }
                merged[record.note_id] = record
            }
            guard merged.count <= 1000,
                  try JSONEncoder().encode(Array(merged.values)).count <= 2_000_000 else { throw NotebookFailure.full }
            if !page.has_more { records = merged; return }
            after = page.next_after
        }
        throw NotebookFailure.full
    }
}
